security/fma-contract-version-upgrade.md

```diff
@@ -0,0 +1,122 @@
+| Created at          | 2024-08-20                         |
+| Needs Approval From | _Security Reviewer Name_           |
+| Other Reviewers     | _Reviewer Name 1, Reviewer Name 2_ |
+| Status              | Draft                              |
+
+> [!NOTE]
+> 📢 Remember:
+>
+> - The single approver in the “Need Approval From” must be from the Security team.
+> - Maintain the “Status” property accordingly.
+
+> [!TIP]
+> Guidelines for writing a good analysis, and what the reviewer will look for:
+>
+> - Show your work: Include steps and tools for each conclusion.
+> - Completeness of risks considered.
+> - Include both implementation and operational failure modes
+> - Provide references to support the reviewer.
+> - The size of the document will likely be proportional to the project's complexity.
+> - The ultimate goal of this document is to identify action items to improve the security of the project. The FMA review process can be accelerated by proactively identifying action items during the writing process.
+
+## Introduction
+
+This document covers solidity version upgrade of all OP Stack smart contracts from 0.8.15 code to 0.8.25.
+
+Below are references for this project:
+
+- [Issue 11527: Smart Contracts: Update all 0.8.15 code to 0.8.25](https://github.com/ethereum-optimism/optimism/issues/11527).
+
+## Failure Modes and Recovery Paths
+
+### Missing side effect on `.selector` access bug (Bug fixed after 0.8.15)
+
+- **Description:**
+  - Introduced: v0.6.2
+  - Fixed: v0.8.21
+  - Explanation: The legacy compiler assumes that conditionals used to determine which selector is used/removed have no side effects. So if the result of the conditional is known at compile time the legacy compiler will hardcode the corresponding selector (since selectors are known at compile time). This bug does no exist when the conditional is used with other types of constants.
+- **Risk Assessment:**
+  - Severity: High (if present)
+  - Likelihood: Low (and mostly present in anti-patterns like using conditionals that have side effects)
+- **Mitigations:** Using a solidity version above 0.8.20, if lower versions must be used, conditionals should have no side effects.
+- **Detection:** Confidence on the absence of this bug can be stronger with robust unit tests that check that all side effects of a call happens and correctly.
+- **Recovery Path(s)**: Redeployment and upgrade of the affected contracts
+
+### Storage write removal before conditional termination (Bug fixed after 0.8.15)
+
+- **Description:**
+
+  - Introduced: v0.8.13
+  - Fixed: v0.8.17
+  - Explanation: Incorrect assumption that a storage write is redundant so the optimizer removes it. But this assumption can be wrong when the developer intended otherwise. A storage write is considered redundant when the storage slot is written to twice within an execution context without that slot being read before the second write or execution unconditionally terminates before the slot is read again.
+
```

Comment:
This is a bug fix. We may have been exposed to the bug but unlikely. The way that this could hurt us is if we depended on the buggy behavior

Reply:
I vaguely recall talking with @maurelian about this a while ago, and think he said we checked our codebase when this was announced? However it's been 2 years since this bug was fixed so we may have introduced this bug at some point after. 

Regardless, agree with @tynes that depending on the buggy behavior is the primary risk, but that is unlikely

Blog post for reference: https://soliditylang.org/blog/2022/09/08/solidity-0.8.17-release-announcement/

Reply:
No, I don't recall doing that check. 

The one that I recall we did a review for was nested [Size Check Bug in Nested Calldata Array ABI-Reencoding](https://soliditylang.org/blog/2022/05/17/calldata-reencode-size-check-bug/).

Reply:
Got it, so we probably should add an action to track verifying this